File: CINE_AI_PLATFORM/apps/api/src/services/shots_service.py
from typing import Any, Dict, Optional
from uuid import uuid4
from datetime import datetime, timezone

from src.storage.json_shots_store import JsonShotsStore


class ShotsService:
    def __init__(self, store: JsonShotsStore):
        self.store = store

    def list_shots(self):
        return self.store.list_shots()

    def get_shot(self, shot_id: str):
        return self.store.get_shot(shot_id)

    def create_shot(self, payload: Dict[str, Any]):
        shot = self._build_shot(payload)
        return self.store.create_shot(shot)

    def replace_shot(self, shot_id: str, payload: Dict[str, Any]):
        current = self.store.get_shot(shot_id)
        if not current:
            return None
        shot = self._build_shot(payload, existing=current)
        return self.store.replace_shot(shot_id, shot)

    def patch_shot(self, shot_id: str, patch_data: Dict[str, Any]):
        current = self.store.get_shot(shot_id)
        if not current:
            return None
        shot = self._build_shot({**current, **patch_data}, existing=current)
        return self.store.replace_shot(shot_id, shot)

    def delete_shot(self, shot_id: str):
        return self.store.delete_shot(shot_id)

    def _now_iso(self) -> str:
        return datetime.now(timezone.utc).isoformat()

    def _normalize_text(self, value: Optional[str]) -> Optional[str]:
        if isinstance(value, str):
            return value.strip()
        return value

    def _build_shot(self, payload: Dict[str, Any], existing: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        base = dict(existing or {})
        shot = {**base, **payload}

        shot["id"] = shot.get("id") or str(uuid4())
        shot["title"] = self._normalize_text(shot.get("title"))
        shot["prompt"] = self._normalize_text(shot.get("prompt"))
        shot["raw_prompt"] = self._normalize_text(shot.get("raw_prompt"))
        shot["negative_prompt"] = self._normalize_text(shot.get("negative_prompt"))
        shot["camera_preset"] = self._normalize_text(shot.get("camera_preset"))
        shot["nominal_ratio"] = self._normalize_text(shot.get("nominal_ratio"))
        shot["scene_id"] = self._normalize_text(shot.get("scene_id"))
        shot["sequence_id"] = self._normalize_text(shot.get("sequence_id"))
        shot["status"] = self._normalize_text(shot.get("status"))

        shot["tags"] = shot.get("tags") if isinstance(shot.get("tags"), list) else []
        shot["references"] = shot.get("references") if isinstance(shot.get("references"), list) else []
        shot["layers"] = shot.get("layers") if isinstance(shot.get("layers"), list) else []
        shot["render_inputs"] = shot.get("render_inputs") if isinstance(shot.get("render_inputs"), dict) else {}
        shot["structured_prompt"] = shot.get("structured_prompt") if isinstance(shot.get("structured_prompt"), dict) else {}
        shot["metadata"] = shot.get("metadata") if isinstance(shot.get("metadata"), dict) else {}

        shot["updated_at"] = self._now_iso()
        if not existing:
            shot["created_at"] = shot["updated_at"]
        else:
            shot["created_at"] = existing.get("created_at", shot["updated_at"])

        return shot

File: CINE_AI_PLATFORM/apps/api/src/services/shots.service.js
import crypto from "node:crypto";
import {
  createShot as createShotInRepo,
  deleteShot as deleteShotInRepo,
  getAllShots,
  getShotById,
  updateShot as updateShotInRepo,
} from "../repositories/shots.repository.js";

function normalizeString(value) {
  return typeof value === "string" ? value.trim() : value;
}

function normalizeArray(value) {
  return Array.isArray(value) ? value : [];
}

function normalizeShotPayload(payload = {}, existingShot = null) {
  const base = existingShot ? { ...existingShot } : {};

  const next = {
    ...base,
    ...payload,
  };

  if (!next.id) {
    next.id = crypto.randomUUID();
  }

  if ("title" in next) next.title = normalizeString(next.title);
  if ("prompt" in next) next.prompt = normalizeString(next.prompt);
  if ("raw_prompt" in next) next.raw_prompt = normalizeString(next.raw_prompt);
  if ("negative_prompt" in next) next.negative_prompt = normalizeString(next.negative_prompt);
  if ("camera_preset" in next) next.camera_preset = normalizeString(next.camera_preset);
  if ("nominal_ratio" in next) next.nominal_ratio = normalizeString(next.nominal_ratio);
  if ("scene_id" in next) next.scene_id = normalizeString(next.scene_id);
  if ("sequence_id" in next) next.sequence_id = normalizeString(next.sequence_id);
  if ("status" in next) next.status = normalizeString(next.status);

  if ("tags" in next) next.tags = normalizeArray(next.tags);
  if ("references" in next) next.references = normalizeArray(next.references);
  if ("layers" in next) next.layers = normalizeArray(next.layers);
  if ("render_inputs" in next) next.render_inputs = next.render_inputs ?? {};
  if ("structured_prompt" in next) next.structured_prompt = next.structured_prompt ?? {};
  if ("metadata" in next) next.metadata = next.metadata ?? {};

  next.updated_at = new Date().toISOString();
  if (!existingShot) {
    next.created_at = next.created_at ?? next.updated_at;
  }

  return next;
}

export async function listShots() {
  return getAllShots();
}

export async function findShotById(id) {
  return getShotById(id);
}

export async function createShot(payload) {
  const shot = normalizeShotPayload(payload);
  return createShotInRepo(shot);
}

export async function updateShot(id, payload) {
  return updateShotInRepo(id, async (existingShot) => {
    if (!existingShot) {
      return null;
    }
    return normalizeShotPayload(payload, existingShot);
  });
}

export async function patchShot(id, patch) {
  return updateShotInRepo(id, async (existingShot) => {
    if (!existingShot) {
      return null;
    }
    return normalizeShotPayload({ ...existingShot, ...patch }, existingShot);
  });
}

export async function removeShot(id) {
  return deleteShotInRepo(id);
}